src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,155 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] DEFAULT_ELEMENTS = {};
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = new Object[DEFAULT_CAPACITY];
+    }
+
+    @SuppressWarnings("unchecked")
+    T elements(int index) {
+        return (T) elements[index];
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index, size);
+        int s;
+        Object[] elementData;
+        if ((s = size) == (elementData = this.elements).length) {
+            elementData = grow(size + 1);
+        }
+        if (index == 0) {
+            System.arraycopy(elementData, 0, elementData, 1, s);
+        } else {
+            System.arraycopy(elementData, 0, elementData, 0, index);
+            System.arraycopy(elementData, index, elementData, index + 1, s - index);
+        }
+        elementData[index] = value;
+        size = s + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            return;
+        }
+        for (int i = 0; i < ((ArrayList<T>) list).size; i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, (size == 0) ? 0 : (size - 1));
+        return elements(index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index, (size == 0) ? 0 : (size - 1));
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, (size == 0) ? 0 : (size - 1));
+        final Object[] es = elements;
+
+        @SuppressWarnings("unchecked") T oldValue = (T) es[index];
+        final int newSize;
+        if ((newSize = size - 1) > index) {
+            System.arraycopy(es, index + 1, es, index, newSize - index);
+        }
+        es[size = newSize] = null;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        Object[] es = elements;
+
+        for (int i = 0; i < this.size; i++) {
+
+            if (equal(element, (T) es[i])) {
+                remove(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException("There is no such element.");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Object[] grow(int minCapacity) {
```

Comment:
```suggestion
    private void grow() {
```

Reply:
you shouldn't pass min capacity in this implementation, you should increase the length of your array every time multiplying on 1.5